src/mempool/mempool.rs

```diff
@@ -26,14 +30,93 @@ use crate::{
 };
 
 use super::routing::{MessageRouter, ShardRouter};
+use governor::{Quota, RateLimiter};
+use moka::sync::{Cache, CacheBuilder};
+use std::num::NonZeroU32;
+use thiserror::Error;
 use tracing::{error, warn};
 
+type DirectRateLimiter = RateLimiter<NotKeyed, InMemoryState, QuantaClock>;
+
+#[derive(Error, Debug)]
+pub enum RateLimitsError {
+    #[error("no storage")]
+    NoStorage,
+}
+
+pub struct RateLimitsConfig {
+    pub time_to_idle: Duration,
+    pub max_capacity: usize,
+}
+
+impl Default for RateLimitsConfig {
+    fn default() -> Self {
+        Self {
+            time_to_idle: Duration::from_secs(60 * 2 * 2),
+            // Make time to idle 2x the rate limit window so it's ok if out of sync
+            max_capacity: 10_000,
+        }
+    }
+}
+
+pub struct RateLimits {
+    shard_stores: HashMap<u32, Stores>,
+    rate_limits_by_fid: Cache<u64, Arc<DirectRateLimiter>>,
+}
+
+impl RateLimits {
+    pub fn new(shard_stores: HashMap<u32, Stores>, config: RateLimitsConfig) -> Self {
+        RateLimits {
+            shard_stores,
+            rate_limits_by_fid: CacheBuilder::new(config.max_capacity)
+                .time_to_idle(config.time_to_idle)
+                .build(),
+        }
+    }
+
+    fn invalidate_rate_limiter_for_fid(&mut self, fid: u64) {
+        self.rate_limits_by_fid.invalidate(&fid);
+    }
+
+    fn get_rate_limiter_for_fid(
+        &mut self,
+        shard_id: u32,
+        fid: u64,
+    ) -> Result<Arc<DirectRateLimiter>, Arc<Box<dyn std::error::Error + Send + Sync>>> {
+        self.rate_limits_by_fid.get_or_try_insert_with(fid, || {
+            let stores = self.shard_stores.get(&shard_id).unwrap();
+            let storage_limits = stores.get_storage_limits(fid).unwrap();
+            let storage_allowance: u32 = storage_limits
+                .limits
+                .iter()
+                .map(|limits| limits.limit as u32)
+                .sum();
+            if storage_allowance == 0 {
+                Err(Box::new(RateLimitsError::NoStorage))
+            } else {
+                Ok(Arc::new(RateLimiter::direct(Quota::per_hour(
```

Comment:
Add a comment here about updating the time_to_idle if we change the quota duration